docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+   reconnects, any events that were published during the period between the
+   disconnect and re-connect and re-subscription will be silently lost, and the
+   client has no way to discover that it missed some relevant messages.
+
+3. **No way to replay old events:** Even if a client knew it had missed some
+   events (due to a disconnection, for example), the API provides no way for
+   the client to "play back" events it may have missed.
+
+4. **Large response sizes:** Some event data can be quite large, and there can
+   be substantial duplication across items. The API allows the client to select
+   _which_ events are reported, but has no way to control which parts of a
+   matching event it wishes to receive.
+
+   This can be costly on the server (which has to marshal those data into
+   JSON), the network, and the client (which has to unmarshal the result and
+   then pick through for the components that are relevant to it).
+
+   Besides being inefficient, this also contributes to some of the persistent
+   connection issues mentioned above, e.g., filling up the websocket write
+   buffer and forcing the server to queue potentially several copies of a large
+   value in memory.
+
+5. **Client identity is tied to network address:** The Tendermint event API
+   identifies each subscriber by a (Client ID, Query) pair. In the RPC service,
+   the query is provided by the client, but the client ID is set to the TCP
+   address of the client (typically "host:port" or "ip:port").
+
+   This means that even if the server did _not_ drop subscriptions immediately
+   when the websocket connection is closed, a client may not be able to
+   reattach to its existing subscription.  Dialing a new connection is likely
+   to result in a different port (and, depending on their own proxy setup,
+   possibly a different public IP).
+
+   In isolation, this problem would be easy to work around with a new
+   subscription parameter, but it would require several other changes to the
+   handling of event subscriptions for that workaround to become useful.
+
+---
+## Decision
+
+(pending)
+
+To address the described problems, we will:
+
+1. Introduce a new API for event subscription to the Tendermint RPC service.
+   The proposed API is described in [Detailed Design](#detailed-design) below.
+
+2. This new API will target the Tendermint v0.36 release, during which the
+   current ("streaming") API will remain available as-is, but deprecated.
+
+3. The streaming API will be entirely removed in release v0.37, which will
+   require all users of event subscription to switch to the new API.
+
+> **Point for discussion:** Given that we already have ABCI++ and PBTS for
+> v0.36, it would be reasonable to suggest postponing this work until v0.37 and
+> its successor instead. The reason I proposed v0.36 for the first step is that
+> the proposed design can be added _alongside_ the existing implementation, so
+> it would not break existing usage.
```

Comment:
The proposed design isn't entirely trivial, but it's not too complicated. My rough approach would be to use the existing pubsub subscriber interface to power the new event log, and then the rest of the logic hangs off the environment.

If we do this, then after we've dropped the old subscriber interface we could rip out most of the pubsub package, as it would no longer be necessary except for tests. That would mean, among other things, that we would no longer need any special queueing logic.

Reply:
I updated this language in light of your comments above, and our offline discussion.